### PR TITLE
Update the CUDA implementation of CudaLuDecomposition with the CUDA matrix class

### PR DESCRIPTION
**Pull request by sjsprecious**
This PR updates the input type for the CUDA implementation of `LuDecompose` function with the `CudaSparseMatrix` class.

All the 42 tests all passed on Derecho's A100 GPU by using `nvhpc/23.7`.

### CODE REVIEW

**Review by mattldawson — APPROVED** [review]
very nice!

File: test/unit/solver/test_cuda_lu_decomposition.cpp [hunk 1]
```diff
@@ -49,76 +50,84 @@ void check_results(
...
+  std::vector<double> cpu_L_vector = cpu_LU.first.AsVector();
+  std::vector<double> cpu_U_vector = cpu_LU.second.AsVector();
+  for (int i = 0; i < L_size; ++i)
+  {
+    EXPECT_EQ(gpu_L_vector[i], cpu_L_vector[i]);
+  };
+  for (int j = 0; j < U_size; ++j)
+  {
+    EXPECT_EQ(gpu_U_vector[j], cpu_U_vector[j]);
+  };
```
Comment (mwaxmonsky) [inline test/unit/solver/test_cuda_lu_decomposition.cpp:106]: Should these be `EXPECT_DOUBLE_EQ` or `EXPECT_NEAR` since they are technically `double`s?

**Review by mwaxmonsky — APPROVED** [review]
Looks great @sjsprecious!  Nothing worth holding it up, just a minor clarification if needed.

File: test/unit/solver/test_cuda_lu_decomposition.cpp (hunk 1 above)
Comment (sjsprecious) [reply test/unit/solver/test_cuda_lu_decomposition.cpp]: Thanks. Revised as suggested.

**Pull request merged**